pilot/pkg/serviceregistry/kube/controller/serviceimportcache.go

```diff
@@ -189,48 +179,78 @@ func (ic *serviceImportCacheImpl) onServiceImportEvent(obj interface{}, event mo
 		// The service already existed. Treat it as an update.
 		event = model.EventUpdate
 
-		// Update the VIPs
-		mcsService.ClusterVIPs.SetAddressesFor(ic.Cluster(), ips)
-		needsFullPush = true
+		if ic.updateIPs(mcsService, ips) {
+			needsFullPush = true
+		}
 	}
 
-	ic.addOrUpdateService(nil, mcsService, event)
+	// Always force a rebuild of the endpoint cache in case this import caused
+	// a change to the discoverability policy.
+	ic.addOrUpdateService(nil, mcsService, event, true)
 
 	if needsFullPush {
-		pushReq := &model.PushRequest{
-			Full: true,
-			ConfigsUpdated: map[model.ConfigKey]struct{}{{
-				Kind:      gvk.ServiceEntry,
-				Name:      mcsHost.String(),
-				Namespace: si.GetNamespace(),
-			}: {}},
-			Reason: []model.TriggerReason{model.ServiceUpdate},
-		}
-		ic.opts.XDSUpdater.ConfigUpdate(pushReq)
+		ic.doFullPush(mcsHost, si.GetNamespace())
 	}
 
 	return nil
 }
 
+func (ic *serviceImportCacheImpl) updateIPs(mcsService *model.Service, ips []string) (updated bool) {
+	prevIPs := mcsService.ClusterVIPs.GetAddressesFor(ic.Cluster())
+	if !util.StringSliceEqual(prevIPs, ips) {
+		// Update the VIPs
+		mcsService.ClusterVIPs.SetAddressesFor(ic.Cluster(), ips)
+		updated = true
+	}
+	return
+}
+
+func (ic *serviceImportCacheImpl) doFullPush(mcsHost host.Name, ns string) {
+	pushReq := &model.PushRequest{
+		Full: true,
+		ConfigsUpdated: map[model.ConfigKey]struct{}{{
+			Kind:      gvk.ServiceEntry,
+			Name:      mcsHost.String(),
+			Namespace: ns,
+		}: {}},
+		Reason: []model.TriggerReason{model.ServiceUpdate},
+	}
+	ic.opts.XDSUpdater.ConfigUpdate(pushReq)
+}
+
+// GetServiceImportIPs returns the list of ClusterSet IPs for the ServiceImport.
+// Exported for testing only.
+func GetServiceImportIPs(si *unstructured.Unstructured) []string {
+	var ips []string
+	if spec, ok := si.Object["spec"].(map[string]interface{}); ok {
+		if rawIPs, ok := spec["ips"].([]interface{}); ok {
+			for _, rawIP := range rawIPs {
+				ip := rawIP.(string)
+				if net.ParseIP(ip) != nil {
+					ips = append(ips, ip)
+				}
+			}
+		}
+	}
+	sort.Strings(ips)
+	return ips
+}
+
+// genMCSService generates an MCS service based on the given real k8s service. The llist of vips must be non-empty.
 func (ic *serviceImportCacheImpl) genMCSService(realService *model.Service, mcsHost host.Name, vips []string) *model.Service {
 	mcsService := realService.DeepCopy()
 	mcsService.Hostname = mcsHost
-
-	if len(vips) > 0 {
-		mcsService.DefaultAddress = vips[0]
-		mcsService.ClusterVIPs.SetAddresses(map[cluster.ID][]string{
-			ic.Cluster(): vips,
-		})
-	} else {
-		mcsService.DefaultAddress = ""
-		mcsService.ClusterVIPs.SetAddresses(nil)
+	mcsService.DefaultAddress = vips[0]
```

Comment:
Is there a validator that would reject a CR with 0 IPs?

Reply:
This is actually fine ... the assignment of the IP may happen after the ServiceImport is created.